src/client/scripts/esm/chess/logic/boardchanges.ts

```
@@ -0,0 +1,294 @@
+		"move": returnPiece,
+		"capture": uncapturePiece,
+	}
+};
+
+// All queue functions queue a change to the board.
+// They add to a changelist which is then executed using a set of changefuncs, see above.
+
+/**
+ * Queues a move with catpure
+ * Need to differentiate this from move so animations can work and so that royal capture can be recognised
+ * @param changes
+ * @param piece The piece moved. Its coords are used as starting coords
+ * @param endCoords 
+ * @param capturedPiece The piece captured
+ */
+function queueCapture(changes: Array<Change>, piece: Piece, endCoords: Coords, capturedPiece: Piece) {
+	changes.push({action: 'capture', piece: piece, endCoords: endCoords, capturedPiece: capturedPiece});
+	return changes;
+}
+
+/**
+ * Queues the addition of a piece to the board
+ * @param changes 
+ * @param piece the piece to add
+ * the pieces index is optional and will get assigned one if none is present
+ */
+function queueAddPiece(changes: Array<Change>, piece: Piece) {
+	changes.push({action: 'add', piece: piece});
+	return changes;
+};
+
+/**
+ * Queues the removal of a piece
+ * @param changes 
+ * @param piece 
+ */
+function queueDeletePiece(changes: Array<Change>, piece: Piece) {
+	changes.push({action: 'delete', piece: piece});
+	return changes;
+}
+
+/**
+ * Moves a piece without capture
+ * @param changes 
+ * @param piece The piece moved. Its coords are used as starting coords
+ * @param endCoords 
+ */
+function queueMovePiece(changes: Array<Change>, piece: Piece, endCoords: Coords) {
+	changes.push({action: 'move', piece: piece, endCoords: endCoords});
+	return changes;
+}
+
+/**
+ * Apply changes in changelist according to changefuncs
+ * @param gamefile the gamefile
+ * @param changes the changes to apply
+ * @param funcs the object contain change funcs
+ */
+function applyChanges(gamefile: gamefile, changes: Array<Change>, funcs: ActionList<genericChangeFunc>) {
+	for (const c of changes) {
+		if (!(c.action in funcs)) continue;
```

Would this be a source of a bug if the change type has a typo and we don't have a change function for it? Is it better to throw an error if the change type is unrecognized?

---

If you look at graphicalchanges, sometimes we don't want to do anything with a change. Also ts will error if it doesn't match any change name.

---

If typescript will error if you create a change with an invalid action, then by that logic we should have zero changes with an invalid action. I'm just going to throw an error if this ever ocurrs, I mean it might happen if we try creating a change in a javascript file, not typescript, then ts wouldn't warn us the change action was invalid.